src/tests/tribol_enzyme_poly_intersect.cpp

```diff
@@ -0,0 +1,308 @@
+    RealT dxidx2[16 * 8];
+    for ( int i{ 0 }; i < 16 * 8; ++i ) {
+      dxidx1[i] = 0.0;
+      dxidx2[i] = 0.0;
+    }
+
+    RealT xi_base[16];
+    for ( int i{ 0 }; i < 16; ++i ) {
+      xi_base[i] = xi[i];
+    }
+
+    for ( int i{ 0 }; i < 4; ++i ) {
+      x_dot[i] = 1.0;
+      __enzyme_fwddiff<void>( (void*)Intersection2DPolygonEnzyme, x1, x_dot, x1 + 4, zeros, 4, x2, zeros, x2 + 4, zeros,
+                              4, pos_tol, d_pos_tol, len_tol, d_len_tol, xi, dxidx1 + 16 * i, xi + 8,
+                              dxidx1 + 16 * i + 8, &num_poly_verts, &d_num_poly_verts );
+      __enzyme_fwddiff<void>( (void*)Intersection2DPolygonEnzyme, x1, zeros, x1 + 4, x_dot, 4, x2, zeros, x2 + 4, zeros,
+                              4, pos_tol, d_pos_tol, len_tol, d_len_tol, xi, dxidx1 + 16 * ( 4 + i ), xi + 8,
+                              dxidx1 + 16 * ( 4 + i ) + 8, &num_poly_verts, &d_num_poly_verts );
+      __enzyme_fwddiff<void>( (void*)Intersection2DPolygonEnzyme, x1, zeros, x1 + 4, zeros, 4, x2, x_dot, x2 + 4, zeros,
+                              4, pos_tol, d_pos_tol, len_tol, d_len_tol, xi, dxidx2 + 16 * i, xi + 8,
+                              dxidx2 + 16 * i + 8, &num_poly_verts, &d_num_poly_verts );
+      __enzyme_fwddiff<void>( (void*)Intersection2DPolygonEnzyme, x1, zeros, x1 + 4, zeros, 4, x2, zeros, x2 + 4, x_dot,
+                              4, pos_tol, d_pos_tol, len_tol, d_len_tol, xi, dxidx2 + 16 * ( 4 + i ), xi + 8,
+                              dxidx2 + 16 * ( 4 + i ) + 8, &num_poly_verts, &d_num_poly_verts );
+      x_dot[i] = 0.0;
+    }
+
+    std::cout << "dxi/dx1 nonzero values:" << std::endl;
+    for ( int j{ 0 }; j < 8; ++j ) {
+      for ( int i{ 0 }; i < 16; ++i ) {
+        auto idx = j * 16 + i;
+        if ( std::abs( dxidx1[idx] ) > 1.0e-15 ) {
+          std::cout << "  (" << i << ", " << j << ") = " << dxidx1[idx] << std::endl;
+        }
+      }
+    }
+
+    std::cout << "dxi/dx2 nonzero values:" << std::endl;
+    for ( int j{ 0 }; j < 8; ++j ) {
+      for ( int i{ 0 }; i < 16; ++i ) {
+        auto idx = j * 16 + i;
+        if ( std::abs( dxidx2[idx] ) > 1.0e-15 ) {
+          std::cout << "  (" << i << ", " << j << ") = " << dxidx2[idx] << std::endl;
+        }
+      }
+    }
+
+    RealT dxidx1_fd[16 * 8];
+    RealT dxidx2_fd[16 * 8];
+    // row 1 assumes x2 is inside x1; row 2 assumes x1 is inside x2
+    RealT x_sgn1[8] = { 1.0, -1.0, -1.0, 1.0, -1.0, 1.0, 1.0, -1.0 };
+    RealT y_sgn1[8] = { 1.0, 1.0, -1.0, -1.0, -1.0, -1.0, 1.0, 1.0 };
+    RealT x_sgn2[8] = { -1.0, 1.0, 1.0, -1.0, 1.0, -1.0, -1.0, 1.0 };
+    RealT y_sgn2[8] = { -1.0, -1.0, 1.0, 1.0, 1.0, 1.0, -1.0, -1.0 };
+    for ( int j{ 0 }; j < 4; ++j ) {
+      x1[j] += x_sgn1[4 * stencil_dir[j] + j] * delta_;
+      for ( int i{ 0 }; i < 16; ++i ) {
+        xi[i] = 0.0;
+      }
+      Intersection2DPolygon( x1, x1 + 4, 4, x2, x2 + 4, 4, pos_tol, len_tol, xi, xi + 8, num_poly_verts, area, true,
```

Comment:
This is a second call to Intersection2DPolygon in the same routine. Can you clarify what the intended calculation is here within a loop over "4"?

Reply:
comment added